### PR TITLE
Replace ktlint with detekt

### PR DESCRIPTION
**Pull request by 0nko**
It seems like there is an issue with `ktlint`, which is causing the builds to fail (and I can't even run it locally). Since we're using `detekt` in other projects, I've `ktlint` with it and generated the baseline.

### CODE REVIEW

**Review by ParaskP7 — COMMENTED** [review]
👋 @0nko !

I have reviewed and tested this PR as per the instructions, everything works as expected, thanks for migrating `KtLint` to `Detekt`! 🌟 🌟 🌟

> It seems like there is an issue with ktlint, which is causing the builds to fail (and I can't even run it locally). Since we're using detekt in other projects, I've ktlint with it and generated the baseline.

From the looks of it `KtLint` doesn't run locally for you probably because you are using a `JDK`  greater than `11`. For example, I am too on `openjdk 17.0.5 2022-10-18` and when running `./gradlew ktlint` locally I get the below failure:

<details>
    <summary>KtLint Failure (JDK 17)</summary>

```
> Task :mediapicker:ktlint FAILED
Exception in thread "main" java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: Could not initialize class org.jetbrains.kotlin.com.intellij.pom.java.LanguageLevel
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at com.pinterest.ktlint.KtlintCommandLine.parallel(Main.kt:575)
        at com.pinterest.ktlint.KtlintCommandLine.parallel$default(Main.kt:547)
        at com.pinterest.ktlint.KtlintCommandLine.lintFiles(Main.kt:298)
        at com.pinterest.ktlint.KtlintCommandLine.run(Main.kt:262)
        at com.pinterest.ktlint.Main.main(Main.kt:70)
Caused by: java.lang.NoClassDefFoundError: Could not initialize class org.jetbrains.kotlin.com.intellij.pom.java.LanguageLevel
        at org.jetbrains.kotlin.com.intellij.core.CoreLanguageLevelProjectExtension.<init>(CoreLanguageLevelProjectExtension.java:26)
        at org.jetbrains.kotlin.com.intellij.core.JavaCoreProjectEnvironment.<init>(JavaCoreProjectEnvironment.java:42)
        at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreProjectEnvironment.<init>(KotlinCoreProjectEnvironment.kt:26)
        at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment$ProjectEnvironment.<init>(KotlinCoreEnvironment.kt:121)
        at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment$Companion.createForProduction(KotlinCoreEnvironment.kt:425)
        at com.pinterest.ktlint.core.internal.KotlinPsiFileFactory.initializePsiFileFactory(KotlinPsiFileFactory.kt:74)
        at com.pinterest.ktlint.core.internal.KotlinPsiFileFactory.acquirePsiFileFactory$ktlint_core(KotlinPsiFileFactory.kt:39)
        at com.pinterest.ktlint.core.KtLint.lint(KtLint.kt:149)
        at com.pinterest.ktlint.internal.FileUtilsKt.lintFile(FileUtils.kt:152)
        at com.pinterest.ktlint.KtlintCommandLine.process(Main.kt:395)
        at com.pinterest.ktlint.KtlintCommandLine.access$process(Main.kt:89)
        at com.pinterest.ktlint.KtlintCommandLine$lintFiles$3$1.call(Main.kt:289)
        at com.pinterest.ktlint.KtlintCommandLine$lintFiles$3$1.call(Main.kt:288)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)

FAILURE: Build failed with an exception.
```

</details> 

> I've migrated the MediaPicker from ktlint to detekt but I'm having some issues getting the buildkite pipeline working.

Yes, I saw [built#183](https://buildkite.com/automattic/wordpress-mediapicker-android/builds/183#_) failing, but I am also seeing that [built#185](https://buildkite.com/automattic/wordpress-mediapicker-android/builds/185#_) is nopw green, and thus, this PR is having all the CI checks as 🟢 and ready to be merged. Did you do something to resolve this problem you were having since yesterday? 🤔 

-----

Below, I have left few questions (❓) and suggestions (💡) for you to consider. Let's discuss them first before proceeding with merging this to `trunk`. 🙏 

1. Question (❓): I noticed you update Kotlin to `1.7.20`, was that somehow necessary? I am asking that because it might be better for use to make this update in a separate PR and isolate this `KtLint-to-Detekt` migration from any extra such changes. Wdyt? 🤔
2. Suggestion (💡): I understand that you copy-pasted the WCAndroid [custom](https://github.com/woocommerce/woocommerce-android/blob/trunk/build.gradle#L49-L65) Detekt configuration and just added a new `detektGenerateBaseline` task. Please consider not going that path and use the default Detekt configuration, just like it was done most recently on FluxC (see [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/build.gradle#L51-L59), [PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2482) for reference). WPAndroid is also using the default Detekt configuration (see [here](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/build.gradle#L142-L152)) and I think it is better to align that configuration across repos. I remember @kidinov, @wzieba and me discussing about that somewhere and that fact that this [custom](https://github.com/woocommerce/woocommerce-android/blob/trunk/build.gradle#L49-L65) Detekt configuration on WCAndroid was, for someone unknown reason, needed at that point, but this is not longer the case, as you see from FluxC and WPAndroid. I am really not trying to increase your effort in merging this, but at the same time I would like to avoid us doing another re-configuration work later on, or having someone else copy-paste this [custom](https://github.com/woocommerce/woocommerce-android/blob/trunk/build.gradle#L49-L65) Detekt configuration for another library/client repo. Let me all know your thoughts on that. 🙏 
3. Suggestion (💡): As I mentioned above, you created another custom `detektGenerateBaseline` task, which is almost a copy-paste of the [custom](https://github.com/woocommerce/woocommerce-android/blob/trunk/build.gradle#L49-L65) main Detekt task, minus the `reports` configuration. I think that is not necessary for the below two reasons:
    - I did the same on this [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/commit/1b720dd710144ccfa6f27423b571a5e8ee170458) I mentioned above, and more specifically this [commit](https://github.com/wordpress-mobile/WordPress-FluxC-Android/commit/1b720dd710144ccfa6f27423b571a5e8ee170458), where I didn't have add that. PS: I have all the details within the commit description.
    - Having said the above, and looking at the number of baseline issues, I actually think that it might not be even worth it to create and maintain a baseline at this point. I am saying that because I am seeing 30 baseline issues, some very easy/quick to fix (like `MagicNumber`), while for anything that might take us more than 1' min to fix, we could just suppress that closer to source and avoid having to create and maintain a baseline from the get going. Wdyt? 🤔 

I know the above are lots to digest Ondrej and apologies if I am overwhelming you with extra work on this `KtLint-to-Detekt` migration, but I really think it is worthy it to do this right now and avoid having to think about it in the future, then also use that as a template for such work on all our other libraries. 🙏 

No matter how this would go, I am here to support you going forward anyway I can. 💯

**Comment by ParaskP7** [comment]
👋 @0nko !

FYI: I forget to mention above that this [ktlint.gradle](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/blob/trunk/ktlint.gradle) can/should be also removed as part of this PR.

**Comment by 0nko** [comment]
@ParaskP7 Thank you for the thorough review! :)

> Did you do something to resolve this problem you were having since yesterday? 🤔

I didn't do anything. yesterday, all the publish tasks were not triggered and I thought I was missing something. There was no output so it wasn't obvious that I should just wait. 🤷 

> I noticed you update Kotlin to 1.7.20, was that somehow necessary? I am asking that because it might be better for use to make this update in a separate PR and isolate this KtLint-to-Detekt migration from any extra such changes. Wdyt? 

You're right, I'll change it back.

> Please consider not going that path and use the default Detekt configuration, just like it was done most recently on FluxC 

Thanks for the suggestion! I wasn't aware I was doing it wrong, I basically just copied what WCAndroid did. I added the `detektGenerateBaseline` because I was trying to figure out how to generate `baseline.xml` but I see that it's not necessary. I'll update the gradle configuration. 👍 

> Having said the above, and looking at the number of baseline issues, I actually think that it might not be even worth it to create and maintain a baseline at this point.

There a couple of `ComplexMethod` or `LongMethod` which would take more time refactoring. I'll fix the easy ones but I'd like to avoid spending too much time on this right now. Is that okay?

**Comment by ParaskP7** [comment]
👋 @0nko and thank you for the update! ❤️ 

> I didn't do anything. yesterday, all the publish tasks were not triggered and I thought I was missing something. There was no output so it wasn't obvious that I should just wait. 🤷

🤷 

> You're right, I'll change it back.

🙇 

> Thanks for the suggestion! I wasn't aware I was doing it wrong, I basically just copied what WCAndroid did. I added the detektGenerateBaseline because I was trying to figure out how to generate baseline.xml but I see that it's not necessary. I'll update the gradle configuration. 👍

👍 🙏 👍 

> There a couple of ComplexMethod or LongMethod which would take more time refactoring. I'll fix the easy ones but I'd like to avoid spending too much time on this right now. Is that okay?

Yes, please, I wasn't suggesting fixing those, instead what I was suggesting is for you to suppress those cases (`ComplexMethod`, `LongMethod` or otherwise) and only fix the very Very VERY easy ones, like (like `MagicNumber`).

Actually, if it is easier on you, you could go ahead and suppress all of them closer to source, no matter if it is anyway easy to fix or not. I am actually suggested this because I think it will be easier for you to do that, instead of dealing with any baseline related configuration, just so that you can avoid spending any more time on it unnecessarily.

Having said that, feel free to totally skip this, if it does take you too much time and I'll follow-up with a PR to do that myself as some point. 💯 

-----

FYI: I'll take another look on all that tomorrow!

**Comment by 0nko** [comment]
Thanks! I've removed the `baseline.xml` file and fixed some more simple violations. I suppressed the rest locally. Ready for another round!

**Comment by kidinov** [comment]
@ParaskP7 

> I remember @kidinov, @wzieba and me discussing about that somewhere and that fact that this [custom](https://github.com/woocommerce/woocommerce-android/blob/trunk/build.gradle#L49-L65) Detekt configuration on WCAndroid was, for someone unknown reason, needed at that point, but this is not longer the case, as you see from FluxC and WPAndroid

I believe I wrote this task because.... not sure 😕, but looking into the comment, I may assume that detekt was checking only the main module by default

**Comment by 0nko** [comment]
@kidinov It's generating the `baseline.xml` for just one, but a regular run checks all of them.

File: build.gradle [hunk 1]
```diff
@@ -7,13 +8,26 @@ plugins {
...
+
+    detekt {
+        config = files("${project.rootDir}/config/detekt/detekt.yml")
+        autoCorrect = false
+        buildUponDefaultConfig = true
+        disableDefaultRuleSets = false
+        ignoreFailures = false
+        parallel = true
+        debug = false
+
```
Comment (ParaskP7) [inline build.gradle]: Minor (🔍): Consider removing this extra empty line here.

**Review by ParaskP7 — APPROVED** [review]
👋 @0nko , this is awesome! ❤️ 

I have reviewed and tested this PR again, everything looks and works as expected, great job on adding Detekt and resolving all those minor quick fix (pending) issues for us! 🌟 🌟 🌟

Feel free to merge it when you feel like it! 🚀

**Comment by 0nko** [comment]
@ParaskP7 Thank you so much! 🥳

**Pull request merged**